ENGG1003 Digital Literacy and Computational Thinking - P 

Course Project 

I declare that the project here submitted is original
<br> 
except for source material explicitly acknowledged,
<br>
and that the same or closely related material has not been
<br> 
previously submitted for another course. 
<br>
I also acknowledge that I am aware of University policy and 
<br>
regulations on honesty in academic work, and of the disciplinary 
<br>
guidelines and procedures applicable to breaches of such 
<br>
policy and regulations, as contained in the website. 

University Guideline on Academic Honesty:
<br> 
https://www.cuhk.edu.hk/policy/academichonesty/

Student Name : PIPATPAJONG Thapakorn

Student ID : 1155184202

Class/Section : ENGG1003EE

Date : 30 Nov 2022 

# ENGG1003 Project

**Please change the path of the files if the file path in jupyter notebook are different from the original settings.**
<br>
**The excel file is the dataset file.**

Firstly, install the necessary packages from python in order to complete the project, in this case, I used the following libraries listed below.

Some of the libraries must be installed first, such as altair, by type **pip install altair** (to use altair) and **pip install openpyxl** (to read excel file).

In [1]:
import pandas as pd
import altair as alt
import numpy as np
import math

ModuleNotFoundError: No module named 'pandas'

### About the project

This project will visualize and analyze about the mortality trends in Hong Kong from 1986 to 2016. The topic about the mortality trends in Hong Kong can be divided into 5 topics, which consist of
1. The mortality rate (crude death and standardize death)
    - The project will compare the different between the crude death and standardize death in Hong Kong from 1986-2016, then visualized and predicted.
2. The mortality rate in different ages and sexes
    - The project compared the mortality rate in Hong Kong and find the trend overtime from 1986-2016 with ages and sexes are the attributes.
3. The causes of death
    - The project will count the different unique cause of death in 1986-2006 and 2006-2016 and then visualized.
4. The expectation of life in different countries
    - This topic will compare and visualize the average lifetime of the people in different countries, and difference sex.
5. Number of deaths in each month.
    - This topic will talk about the number deaths monthly ranging from 1986-2016.

The data are gotten from the Census and Statistics department in Hong Kong. The data will be visualized, modeled, interpreted, analyzed, and predicted in the future thoroughly.


### Data Retrieval

The datasets in this project can be downloaded from the following link: **https://www.censtatd.gov.hk/en/EIndexbySubject.html?scode=150&pcode=B1120017 .** In this downlanding link, the datasets will be the demographic trends in Hong Kong, which consist of many topic about the demographic trends. This project focuses solely on the mortality trends, so only the topic 4 in the datasets will be used.

From the imported libraries above, the data will be read by using pandas by code

**df = pd.read_csv('(file path)')** --> the data retrival will be pointed out in the code by comment

## 1. The mortality rate (Crude death and standardize death)

The crude death is the straightforward way to measure the death rate, the formula (per 1000 people) is

$CDR = 1000\times(\frac{N_{deaths}}{N_{population}})$

$CDR =$ Crude death rates
<br>
$N_{deaths} =$ Number of deaths occurred
<br>
$N_{population} =$ Number of population observed

On the other hand, the standardize rate is the rate that is weighted on the risk of death because the older one get the higher chance of death is increasing. Thus, the formula (per 1000 people) is

$SDR = \sum_{age} \frac{m_{age}p_{age}}{p_{age}}$

$SDR =$ Standardized death rates
<br>
$m_{age} =$ The death rate of the specific $age$
<br>
$p_{age} =$ The number of population observed of the specific $age$

"Glossary:Standardised death rate (SDR)," Eurostat, accessed December 2, 2022, https://ec.europa.eu/eurostat/statistics-explained/index.php?title=Glossary:Standardised_death_rate_(SDR)

The means the crude death rate count the number of death, while standardized death rate is counting by weighting the expected cases higher than the unexpected cases. Thus, the higher the standardized rate is, the more expected death happened (which means less accidents or other kind of immediate deaths.)

### Data retrieval and data cleaning/filtering

In [ ]:
# Data retrieval
df = pd.read_excel(r"Mortality_trends.xlsx", sheet_name = 'CDR_SDR')

# Do some data cleaning and filtering
# Since the population in each year is not equal, thus we are focusing on the crude rate and standardized rate.
df = df[['Yr', 'Crude_death_r', 'Stand_death_r']]
df

Yr  Crude_death_r  Stand_death_r
0   1986            4.7           10.5
1   1987            4.8           10.6
2   1988            4.9           10.6
3   1989            5.1           10.7
4   1990            5.2           10.8
5   1991            5.0           10.1
6   1992            5.3           10.6
7   1993            5.2           10.2
8   1994            5.0            9.5
9   1995            5.1            9.4
10  1996            5.0            9.2
11  1997            4.9            8.9
12  1998            5.0            9.0
13  1999            5.0            8.9
14  2000            5.1            8.7
15  2001            5.0            8.1
16  2002            5.1            8.0
17  2003            5.5            8.4
18  2004            5.4            8.1
19  2005            5.7            8.2
20  2006            5.5            7.5
21  2007            5.7            7.6
22  2008            6.0            7.8
23  2009            5.9            7.3
24  2010            6.0            7.3
25  2011            6.0            7.0
26  2012            6.1            6.9
27  2013            6.0            6.6
28  2014            6.2            6.6
29  2015            6.3            6.5
30  2016            6.4            6.4

### Data summarization

In [ ]:
df.describe()

Yr  Crude_death_r  Stand_death_r
count    31.000000      31.000000      31.000000
mean   2001.000000       5.422581       8.580645
std       9.092121       0.502467       1.435367
min    1986.000000       4.700000       6.400000
25%    1993.500000       5.000000       7.400000
50%    2001.000000       5.200000       8.400000
75%    2008.500000       5.950000       9.800000
max    2016.000000       6.400000      10.800000

From the description above, the standardize death rate are always higher than the crude death rate. However the standard deviation of standardize death rate is higher than crude death rate. Which shows that the standardize death rate is more fluctuated than the crude death rate. The minimum value, 25 percentile, 50 percentile, 75 percentile, maximum value in crude death rate are close to each other compared to standardize death, which agrees with the standard deviation value.

### Data processing

Create a new column named "Type of death" to store the types of death and then use that to visualize data between crude death rate and standardize death rate.

In [ ]:
# Do some more data processing by create a new variable which is type of death so that the data can be visualized by separated the death types.
# we want to plot the rate in both crude death rate and standardize death rate, so we do a few changes in the dataset.
df_2 = df.melt(id_vars=['Yr'], value_vars=['Crude_death_r', 'Stand_death_r'], var_name='Type of death', value_name='Rate (per 1000 population)')
df_2

Yr  Type of death  Rate (per 1000 population)
0   1986  Crude_death_r                         4.7
1   1987  Crude_death_r                         4.8
2   1988  Crude_death_r                         4.9
3   1989  Crude_death_r                         5.1
4   1990  Crude_death_r                         5.2
..   ...            ...                         ...
57  2012  Stand_death_r                         6.9
58  2013  Stand_death_r                         6.6
59  2014  Stand_death_r                         6.6
60  2015  Stand_death_r                         6.5
61  2016  Stand_death_r                         6.4

[62 rows x 3 columns]

### Data visualization and Data modelling

Create the area graph which represents the real value of the death rates, and the line graph which represent the trends of the data (linear regression trends)

In [ ]:
area = alt.Chart(df_2).mark_area(opacity=0.3).encode(
    x = 'Yr',
    y = alt.Y('Rate (per 1000 population)', stack=None),
    color = 'Type of death'
)

base = alt.Chart(df).mark_circle(opacity=0.7).transform_fold(
    fold = ['Crude_death_r', 'Stand_death_r'],
    as_  = ['Type of death', 'Rate (per 1000 population)']
).encode(
    alt.X('Yr'),
    alt.Y('Rate (per 1000 population):Q'),
    alt.Color('Type of death:N')
)

area + base.transform_regression('Yr', 'Rate (per 1000 population)', groupby=['Type of death']).mark_line(size=3)

c:\Users\Thapakorn\AppData\Local\Programs\Python\Python310\lib\site-packages\altair\utils\core.py:317: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  for col_name, dtype in df.dtypes.iteritems():


alt.LayerChart(...)

### Converting the data into equation/Observing the trends

The trend from the graph above seems to be the linear regession graph because the linear graph fit the trends very well. The data will be converted into the equation to make some predictions.

In [ ]:
# from the graph, it seems like the trend is linear
trendline_crude = np.polyfit(df['Yr'], df['Crude_death_r'], 1)
trendline_stand = np.polyfit(df['Yr'], df['Stand_death_r'], 1)

print('The linear equation of the crude death rate is Crude_death_r = Yr *', trendline_crude[0], '%+.14f' % (trendline_crude[1]))
print('The linear equation of the standardiize death rate is Stand_death_r = Yr *', trendline_stand[0], '%+.14f' % (trendline_stand[1]))

The linear equation of the crude death rate is Crude_death_r = Yr * 0.05020161290322541 -95.03084677419274
The linear equation of the standardiize death rate is Stand_death_r = Yr * -0.15536290322580515 +319.46181451612637


### Data interpretation

After observing the trends and getting the linear equation, use the equation to predict the trend. In this project, the values will be predicted up to year 2050.

In [ ]:
# predict the year up to 2050

for i in range(2017, 2051):
    df_pred = pd.DataFrame({'Yr': [i],
                        'Crude_death_r': [i*trendline_crude[0] + trendline_crude[1]],
                        'Stand_death_r': [i*trendline_stand[0] + trendline_stand[1]]})
    df = pd.concat([df, df_pred], ignore_index=True)
df

Yr  Crude_death_r  Stand_death_r
0   1986       4.700000      10.500000
1   1987       4.800000      10.600000
2   1988       4.900000      10.600000
3   1989       5.100000      10.700000
4   1990       5.200000      10.800000
..   ...            ...            ...
60  2046       7.681653       1.589315
61  2047       7.731855       1.433952
62  2048       7.782056       1.278589
63  2049       7.832258       1.123226
64  2050       7.882460       0.967863

[65 rows x 3 columns]

After got the values in dataset, the visualization is the same as the above one

In [ ]:
df_2 = df.melt(id_vars=['Yr'], value_vars=['Crude_death_r', 'Stand_death_r'], var_name='Type of death', value_name='Rate (per 1000 population)')
df_2

area = alt.Chart(df_2).mark_area(opacity=0.3).encode(
    x = 'Yr',
    y = alt.Y('Rate (per 1000 population)', stack=None),
    color = 'Type of death'
)

base = alt.Chart(df).mark_circle(opacity=0.7).transform_fold(
    fold = ['Crude_death_r', 'Stand_death_r'],
    as_  = ['Type of death', 'Rate (per 1000 population)']
).encode(
    alt.X('Yr'),
    alt.Y('Rate (per 1000 population):Q'),
    alt.Color('Type of death:N')
)

area + base.transform_regression('Yr', 'Rate (per 1000 population)', groupby=['Type of death']).mark_line(size=3)

c:\Users\Thapakorn\AppData\Local\Programs\Python\Python310\lib\site-packages\altair\utils\core.py:317: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  for col_name, dtype in df.dtypes.iteritems():


alt.LayerChart(...)

#### Some suggestions

Even though the trends in the first 30 years is the linear regression models, the author thinks that this model can be predicted in a short term. Otherwise, the death rate will be negative, which is completely impossible. In order to correctly predict the death in a long term, the sufficient data are needed, so that the equations of the trends are more precise.

### Conclusion

&emsp;In 1986, the standardized death rate (SDR) was much higher than the crude death rate (CDR) by 173%. However, the trend of SDR is decreasing and the trend of CDR is increasing. Eventually met at the same value (rate = 6.4) in 2016. This can predicted the trends after 2016 that the SDR will be less than the CDR. By prediction using linear regression, in 2050, the CDR will be approximately 7.88, which is 7 times higher than the SDR. From the definition of SDR and CDR, this implies that there will be more unexpected mortality in future (such as accident, extremely bad disease) than the expected incident one (such as death from being old). This is the signal to warn everyone that everyone should be more careful when doing anything.

## 2. The mortality rates in different ages and sexes

### Data retreival and data cleaning/filtering

In [ ]:
df = pd.read_excel(r"Mortality_trends.xlsx", sheet_name = 'Age_Sex_Mortality')

# we want to compute in different ages
df = df[df['Age_gp'] != 'All ages']
df

Yr     Sex   Age_gp  Mor_rate
0    1986    Male        0       7.5
1    1986    Male    1 – 4       0.3
2    1986    Male    5 – 9       0.2
3    1986    Male  10 – 14       0.2
4    1986    Male  15 – 19       0.4
..    ...     ...      ...       ...
274  2016  Female  65 – 69       5.9
275  2016  Female  70 – 74       9.6
276  2016  Female  75 – 79      17.1
277  2016  Female  80 – 84      33.4
278  2016  Female      85+      89.3

[266 rows x 4 columns]

### Data summarization

In [ ]:
df.describe()

Yr    Mor_rate
count   266.00000  266.000000
mean   2001.00000   14.898872
std      10.01885   29.461981
min    1986.00000    0.100000
25%    1991.00000    0.300000
50%    2001.00000    1.850000
75%    2011.00000   10.675000
max    2016.00000  133.700000

This time the data are extremely fluctuated, which can see by looking at the Mortality rate between 75 percentile and the maximum value. The maximum value is almost 13 times higher than the 75 percentile and approximately 70 times higher than the median (50%). Therefore, the data selection are needed.

### Data selection

Observing the data from the data summarization section and the original data, the mortality rate can be ranged from 0.1 to 133.7. The mortality dropped after the infant state, and then grew exponentially after that. Thus, it is safe to use the data at age group = 0 and the data from the older age group, in this project, the age group over 60 years old will be used.

In [ ]:
# focus on the infant and person in the age over 60 group
df_selected = pd.concat([df[df['Age_gp'] == '0'], df[df['Age_gp'] > '60']], ignore_index = True)
df_selected

Yr     Sex   Age_gp  Mor_rate
0   1986    Male        0       7.5
1   1991    Male        0       6.2
2   1996    Male        0       4.3
3   2001    Male        0       2.9
4   2006    Male        0       2.8
..   ...     ...      ...       ...
93  2016  Female  65 – 69       5.9
94  2016  Female  70 – 74       9.6
95  2016  Female  75 – 79      17.1
96  2016  Female  80 – 84      33.4
97  2016  Female      85+      89.3

[98 rows x 4 columns]

### Data visualization

Since there are many variable to visualize, the author chose grouped bar graph to visualize

In [ ]:
alt.Chart(df_selected).mark_bar().encode(
    x = 'Sex',
    y = 'Mor_rate',
    color = 'Age_gp',
    column = 'Yr:N'
)

c:\Users\Thapakorn\AppData\Local\Programs\Python\Python310\lib\site-packages\altair\utils\core.py:317: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  for col_name, dtype in df.dtypes.iteritems():


alt.Chart(...)

### Conclusion

&emsp;From the datasets and the graph, the mortality rate of the infant (age group = 0) is relatively high compared to other baby ages. In the author opinion, this is because the extremely high vulnerability of the infant, with improper caring, the infant could be easily dead. 

&emsp;Overall, the overall mortality rate of female is much less than the male in every age group except for the infant group. The overall mortality rate of the female is approximately 25-30% less than the male.

&emsp;Comparing age-wise, the mortality rate is exponentially grow when getting into older group, which is reasonable. The mortality rate of male in 1986 going from 65.6 (per 1000 population) in the 75-79 age group to 133.7 (per 1000 population) in the 85+ age group, which is around 2 times higher.

## 3. The causes of death

In this topic, there are 2 files to work with.

### Data retreival and Data cleaning/filtering

In [ ]:
df1 = pd.read_excel(r"Mortality_trends.xlsx", sheet_name = 'Causes_A')

# filter some data out
df1 = df1.drop(columns = ['No_reg_deaths'])
df1 = df1[df1['COD'] != 'Total'].reset_index(drop = True)

df1

Yr         Sex                                                COD  \
0   1986        Male                                          Neoplasms   
1   1986        Male                 Diseases of the circulatory system   
2   1986        Male                 Diseases of the respiratory system   
3   1986        Male                               Injury and poisoning   
4   1986        Male               Diseases of the genitourinary system   
..   ...         ...                                                ...   
85  2000  Both sexes                   Diseases of the digestive system   
86  2000  Both sexes  Endocrine, nutritional and metabolic diseases ...   
87  2000  Both sexes                  Infectious and parasitic diseases   
88  2000  Both sexes         Symptoms, signs and ill-defined conditions   
89  2000  Both sexes                                       Other causes   

    Per_deaths  
0         34.0  
1         26.5  
2         16.4  
3          6.9  
4          3.6  
..         ...  
85         4.2  
86         2.7  
87         2.1  
88         1.1  
89         1.9  

[90 rows x 4 columns]

In [ ]:
df2 = pd.read_excel(r"Mortality_trends.xlsx", sheet_name = 'Causes_B')

# filter some data out
df2 = df2.drop(columns = ['No_reg_deaths'])
df2 = df2[df2['COD'] != 'Total'].reset_index(drop = True)

df2

Yr         Sex                                                COD  \
0    2006        Male                                          Neoplasms   
1    2006        Male                 Diseases of the respiratory system   
2    2006        Male                 Diseases of the circulatory system   
3    2006        Male               Diseases of the genitourinary system   
4    2006        Male         External causes of morbidity and mortality   
..    ...         ...                                                ...   
139  2016  Both sexes  Diseases of the musculoskeletal system and con...   
140  2016  Both sexes       Diseases of the skin and subcutaneous tissue   
141  2016  Both sexes  Diseases of the blood and blood-forming organs...   
142  2016  Both sexes  Congenital malformations, deformations and chr...   
143  2016  Both sexes                                       Other causes   

     Per_deaths  
0          35.8  
1          18.7  
2          23.2  
3           3.9  
4           6.0  
..          ...  
139         0.4  
140         0.4  
141         0.2  
142         0.1  
143         0.1  

[144 rows x 4 columns]

### Data summarization

In [ ]:
df1.describe()

Yr  Per_deaths
count    90.000000   90.000000
mean   1994.000000    9.997778
std       5.920826   11.089705
min    1986.000000    0.800000
25%    1986.000000    2.325000
50%    1996.000000    4.150000
75%    2000.000000   16.550000
max    2000.000000   36.400000

In [ ]:
df2.describe()

Yr  Per_deaths
count   144.000000  144.000000
mean   2011.000000    6.246528
std       4.096732    9.665669
min    2006.000000    0.100000
25%    2006.000000    0.400000
50%    2011.000000    2.200000
75%    2016.000000    4.275000
max    2016.000000   35.800000

From the summarization, the number of data in 1986-2000 are less than the ones in 2006-2016, which means there are more causes of death in 2006-2016. The value in other aspects are similar to each other, except the 75 percentile of 1986-2000 data is 4 times higher than ones in 2006-2016. However, the standard deviation of the 1986-2000 version is higher than the 2006-2016 with the close value of maximum value. That means in 1986-2000, the values are fluctuated a lot.

### Data visualization

Every value in a single year is added up to 100%. The author visualized the graph by using horizontal bar graph as shown below.

In [ ]:
df = pd.concat([df1, df2], ignore_index = True)

alt.Chart(df).mark_bar().encode(
    x = 'Per_deaths',
    y = 'Sex',
    color = alt.Color('COD'),
    row = 'Yr'
)

c:\Users\Thapakorn\AppData\Local\Programs\Python\Python310\lib\site-packages\altair\utils\core.py:317: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  for col_name, dtype in df.dtypes.iteritems():


alt.Chart(...)

### Analyze the difference of the data

As the graph suggested, there are 19 unique causes (excluding the category "other cases") of death during 1986-2000 and 2006-2016. Therefore, the author wants to find the unique cases that exclusively have in 1986-2000 period and 2006-2016 period, and the cases that both periods have. 

In [ ]:
# the unique causes of death are not the same (in terms of amount and context)

def count_unique(array):
    for i in range(len(array)):
        for j in range(len(array)-1, i+1, -1):
            if array[i] == array[j] and array[i] != 0 and array[j] != 0:
                array[j] = 0
    
    count = len(array)            
    for i in range(len(array)):
        if array[i] == 0:
            count -= 1
    return count

df1_copy = df1.copy()

print('The number unique cause of death from 1986-2000 is', count_unique(df1_copy['COD']), ', which are')
for i in df1_copy['COD']:
    if i != 0:
        print(i)


The number unique cause of death from 1986-2000 is 10 , which are
Neoplasms
Diseases of the circulatory system
Diseases of the respiratory system
Injury and poisoning
Diseases of the genitourinary system
Diseases of the digestive system
Endocrine, nutritional and metabolic diseases and immunity disorders
Infectious and parasitic diseases
Symptoms, signs and ill-defined conditions
Other causes


C:\Users\Thapakorn\AppData\Local\Temp\ipykernel_3396\1958885029.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  array[j] = 0


In [ ]:
df2_copy = df2.copy()
print('The number unique cause of death from 2006-2016 is', count_unique(df2_copy['COD']), ', which are')
for i in df2_copy['COD']:
    if i != 0:
        print(i)

C:\Users\Thapakorn\AppData\Local\Temp\ipykernel_3396\1958885029.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  array[j] = 0


The number unique cause of death from 2006-2016 is 16 , which are
Neoplasms
Diseases of the respiratory system
Diseases of the circulatory system
Diseases of the genitourinary system
External causes of morbidity and mortality
Diseases of the digestive system
Symptoms, signs and abnormal clinical and laboratory findings, not elsewhere classified
Mental and behavioural disorders
Certain infectious and parasitic diseases
Endocrine, nutritional and metabolic diseases
Diseases of the nervous system
Diseases of the musculoskeletal system and connective tissue
Diseases of the skin and subcutaneous tissue
Diseases of the blood and blood-forming organs and certain disorders involving the immune mechanism
Congenital malformations, deformations and chromosomal abnormalities
Other causes


For the result, the causes that both periods have are (excluding "Other causes")
- Neoplasms
- Diseases of the circulatory system
- Diseases of the respiratory system
- Diseases of the genitourinary system
- Diseases of the digestive system

the causes exclusively in 1986-2000 are
- Injury and poisoning
- Endocrine, nutritional and metabolic diseases and immunity disorders
- Infectious and parasitic diseases
- Symptoms, signs and ill-defined conditions

the causes exclusively in 2006-2016 are
- External causes of morbidity and mortality
- Symptoms, signs and abnormal clinical and laboratory findings, not elsewhere classified
- Mental and behavioural disorders
- Certain infectious and parasitic diseases
- Endocrine, nutritional and metabolic diseases
- Diseases of the nervous system
- Diseases of the musculoskeletal system and connective tissue
- Diseases of the skin and subcutaneous tissue
- Diseases of the blood and blood-forming organs and certain disorders involving the immune mechanism
- Congenital malformations, deformations and chromosomal abnormalities

### Conclusion

&emsp;Despite there are some new unique causes of death happened after 2000, there are some diseases that can be analyzed. Firstly, the most common case is following:
1. Neoplasms                            average percentage of male = 34.63, female = 28.80
2. Diseases of circulatory system       average percentage of male = 23.73, female = 28.58
3. Diseases of the respiratory system   average percentage of male = 19.78, female = 18.18

Those 3 disease combined contribute for 78.14% for male and 75.56% for female, or around 3 out of 4 of all causes happended in Hong Kong.

&emsp;From the trends in the span of 30 years, the Neoplasms seems to be constant, neither decreasing nor increasing. For the circulatory system, the death percentage is slightly decreasing from 28.9% in 1986 to 21.5% in 2016. On the other hand, the respiratory system has the opposite trend of the circulatory system, the trend is increasing from 15.8% in 1986 to 22.8% in 2016. The proportion increased for almost 50% during the 30-year period.

## 4. The expectation of life in different countries

### Data retrieval/ Data cleaning/filtering

In [ ]:
df = pd.read_excel(r"Mortality_trends.xlsx", sheet_name = 'Expectation')

# fliter some irrelevant data
df.drop('Stat_var', inplace = True, axis = 1)
df = df[df['Sex'] != 'Difference (Female – Male)']

# reindexing
df = df.reset_index(drop = True)
df

Yr     Sex      Economy Exp_of_life_birth
0    1986    Male    Hong Kong              74.1
1    1986    Male        Japan              75.2
2    1986    Male    Singapore              72.1
3    1986    Male    Australia              72.8
4    1986    Male      Denmark              71.6
..    ...     ...          ...               ...
259  2016  Female  Netherlands              83.1
260  2016  Female       Norway              84.2
261  2016  Female       Sweden              84.1
262  2016  Female         U.K.              82.9
263  2016  Female       U.S.A.              N.A.

[264 rows x 4 columns]

### Observing the data

There are many countries to display. Thus, the author want to find the number of countries in order to make the appropriate visualizations.

In [ ]:
# Counting the number of country in dataset
df_copy = df.copy()
print('The number of the countries in this dataset is', count_unique(df_copy['Economy']), ', which are')
for i in df_copy['Economy']:
    if i != 0:
        print(i)

C:\Users\Thapakorn\AppData\Local\Temp\ipykernel_3396\1958885029.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  array[j] = 0


The number of the countries in this dataset is 12 , which are
Hong Kong
Japan
Singapore
Australia
Denmark
France
Germany
Netherlands
Norway
Sweden
U.K.
U.S.A.


Since there are 12 countries, the author found that using bubble map visualization is the most appropriate way to visualize the data.

However, before visualizing, there are some unfilled data that need to be replaced.

### Data modeling

Please take a look at the result of this code:

In [ ]:
print(df[df['Exp_of_life_birth'] == '..'])
print(df[df['Exp_of_life_birth'] == 'N.A.'])

      Yr     Sex  Economy Exp_of_life_birth
6   1986    Male  Germany                ..
18  1986  Female  Germany                ..
       Yr     Sex  Economy Exp_of_life_birth
222  2015    Male  Germany              N.A.
234  2015  Female  Germany              N.A.
246  2016    Male  Germany              N.A.
251  2016    Male   U.S.A.              N.A.
258  2016  Female  Germany              N.A.
263  2016  Female   U.S.A.              N.A.


There are some ".." and "N.A." values in the expectation of life column. The author wants to find the trends of the graph by countries and sexes (because male and female mortality trends are different, refer from second topic) by visualize the curve and find the appropriate trends for the graph.

There are only Germany's and U.S.A.'s values missing. 

In [ ]:
# only Germany & U.S.A have unfilled value
# find the trend of the filled values

# do for the Germany first

df_ger = df[df['Economy'] == 'Germany'][df['Exp_of_life_birth'] != '..'][df['Exp_of_life_birth'] != 'N.A.'][df['Sex'] != 'Difference (Female – Male)']

alt.Chart(df_ger).mark_line().encode(
    x = 'Yr:N',
    y = 'Exp_of_life_birth:Q',
    color = 'Sex'
)

C:\Users\Thapakorn\AppData\Local\Temp\ipykernel_3396\1951122765.py:6: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df_ger = df[df['Economy'] == 'Germany'][df['Exp_of_life_birth'] != '..'][df['Exp_of_life_birth'] != 'N.A.'][df['Sex'] != 'Difference (Female – Male)']
c:\Users\Thapakorn\AppData\Local\Programs\Python\Python310\lib\site-packages\altair\utils\core.py:317: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  for col_name, dtype in df.dtypes.iteritems():


alt.Chart(...)

In [ ]:
# This time USA

df_usa = df[df['Economy'] == 'U.S.A.'][df['Exp_of_life_birth'] != '..'][df['Exp_of_life_birth'] != 'N.A.'][df['Sex'] != 'Difference (Female – Male)']

alt.Chart(df_usa).mark_line().encode(
    x = 'Yr:N',
    y = 'Exp_of_life_birth:Q',
    color = 'Sex'
)

C:\Users\Thapakorn\AppData\Local\Temp\ipykernel_3396\2862361121.py:3: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df_usa = df[df['Economy'] == 'U.S.A.'][df['Exp_of_life_birth'] != '..'][df['Exp_of_life_birth'] != 'N.A.'][df['Sex'] != 'Difference (Female – Male)']


alt.Chart(...)

In the author opinion, the trends of both countries share the similarity trends. The graphs are steep in the 1986-2006, and then slightly grow afterwards. Which means the trendlines are logarithmic function. Then fill all the Germany's and U.S.A.'s missing values.

In [ ]:
# For me, it seems both of the countries have trendline with logarithmic function
# starts with Germany

df_ger[['Yr', 'Exp_of_life_birth']] = df_ger[['Yr', 'Exp_of_life_birth']].apply(pd.to_numeric)

# male
df_ger_m = df_ger[df_ger['Sex'] == 'Male']
curve_m = np.polyfit(np.log(df_ger_m['Yr']), df_ger_m['Exp_of_life_birth'], 1)

# the missing value is Yr = 1986 (col 6), 2015 (col 222), 2016 (col 246)
df.iat[6, 3] = np.log(df.iat[6, 0])*curve_m[0] + curve_m[1]
df.iat[222, 3] = np.log(df.iat[222, 0])*curve_m[0] + curve_m[1] 
df.iat[246, 3] = np.log(df.iat[246, 0])*curve_m[0] + curve_m[1]
df[df['Economy'] == 'Germany'][df['Sex'] == 'Male']

C:\Users\Thapakorn\AppData\Local\Temp\ipykernel_3396\16751797.py:14: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df[df['Economy'] == 'Germany'][df['Sex'] == 'Male']


Yr   Sex  Economy Exp_of_life_birth
6    1986  Male  Germany         71.577587
30   1991  Male  Germany              72.9
54   1996  Male  Germany              73.6
78   2001  Male  Germany              75.4
102  2006  Male  Germany              76.9
126  2011  Male  Germany              77.7
150  2012  Male  Germany              77.9
174  2013  Male  Germany              78.1
198  2014  Male  Germany              78.2
222  2015  Male  Germany         78.645568
246  2016  Male  Germany         78.887474

In [ ]:
# female
df_ger_f = df_ger[df_ger['Sex'] == 'Female']
curve_f = np.polyfit(np.log(df_ger_f['Yr']), df_ger_f['Exp_of_life_birth'], 1)

# the missing value is Yr = 1986 (col 18), 2015 (col 234), 2016 (col 258)
df.iat[18, 3] = np.log(df.iat[18, 0])*curve_f[0] + curve_f[1]
df.iat[234, 3] = np.log(df.iat[234, 0])*curve_f[0] + curve_f[1] 
df.iat[258, 3] = np.log(df.iat[258, 0])*curve_f[0] + curve_f[1]
df[df['Economy'] == 'Germany'][df['Sex'] == 'Female']

C:\Users\Thapakorn\AppData\Local\Temp\ipykernel_3396\4208265484.py:9: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df[df['Economy'] == 'Germany'][df['Sex'] == 'Female']


Yr     Sex  Economy Exp_of_life_birth
18   1986  Female  Germany         78.474855
42   1991  Female  Germany              79.3
66   1996  Female  Germany                80
90   2001  Female  Germany              81.2
114  2006  Female  Germany              82.3
138  2011  Female  Germany              82.8
162  2012  Female  Germany              82.9
186  2013  Female  Germany              83.1
210  2014  Female  Germany              83.1
234  2015  Female  Germany         83.468037
258  2016  Female  Germany         83.638931

In [ ]:
# U.S.A.
df_usa[['Yr', 'Exp_of_life_birth']] = df_usa[['Yr', 'Exp_of_life_birth']].apply(pd.to_numeric)

# male
df_usa_m = df_usa[df_usa['Sex'] == 'Male']
curve_um = np.polyfit(np.log(df_usa_m['Yr']), df_usa_m['Exp_of_life_birth'], 1)

# the missing value is Yr = 2016 (Col 251)
df.iat[251, 3] = np.log(df.iat[251, 0])*curve_um[0] + curve_um[1]

# female
df_usa_f = df_usa[df_usa['Sex'] == 'Female']
curve_uf = np.polyfit(np.log(df_usa_f['Yr']), df_usa_f['Exp_of_life_birth'], 1)

# the missing value is Yr = 2016 (Col 263)
df.iat[263, 3] = np.log(df.iat[263, 0])*curve_uf[0] + curve_uf[1]
df[df['Economy'] == 'U.S.A.']

Yr     Sex Economy Exp_of_life_birth
11   1986    Male  U.S.A.              71.2
23   1986  Female  U.S.A.              78.2
35   1991    Male  U.S.A.                72
47   1991  Female  U.S.A.              78.9
59   1996    Male  U.S.A.              73.1
71   1996  Female  U.S.A.              79.1
83   2001    Male  U.S.A.              74.3
95   2001  Female  U.S.A.              79.5
107  2006    Male  U.S.A.              75.2
119  2006  Female  U.S.A.              80.3
131  2011    Male  U.S.A.              76.3
143  2011  Female  U.S.A.              81.1
155  2012    Male  U.S.A.              76.4
167  2012  Female  U.S.A.              81.2
179  2013    Male  U.S.A.              76.4
191  2013  Female  U.S.A.              81.2
203  2014    Male  U.S.A.              76.5
215  2014  Female  U.S.A.              81.3
227  2015    Male  U.S.A.              76.3
239  2015  Female  U.S.A.              81.2
251  2016    Male  U.S.A.         76.970953
263  2016  Female  U.S.A.         81.476268

In [ ]:
df[['Exp_of_life_birth']] = df[['Exp_of_life_birth']].apply(pd.to_numeric)

# now the data is cleaned!
df

Yr     Sex      Economy  Exp_of_life_birth
0    1986    Male    Hong Kong          74.100000
1    1986    Male        Japan          75.200000
2    1986    Male    Singapore          72.100000
3    1986    Male    Australia          72.800000
4    1986    Male      Denmark          71.600000
..    ...     ...          ...                ...
259  2016  Female  Netherlands          83.100000
260  2016  Female       Norway          84.200000
261  2016  Female       Sweden          84.100000
262  2016  Female         U.K.          82.900000
263  2016  Female       U.S.A.          81.476268

[264 rows x 4 columns]

The dataset is eventually cleaned and ready to use

### Data Visualization

From the observation above, the author planned to use the bubble map visualization.

Visualize separately for the male and the female, then combined the graph.

In [ ]:
male = alt.Chart(df[df['Sex'] == 'Male']).mark_circle(opacity = 0.5, color = 'blue').encode(
    x = alt.X('Yr:N'),
    y = alt.Y('Economy:O'),
    size = alt.Size('Exp_of_life_birth', scale = alt.Scale(zero = False))
)

female = alt.Chart(df[df['Sex'] == 'Female']).mark_circle(opacity = 0.5, color = 'red').encode(
    x = alt.X('Yr:N'),
    y = alt.Y('Economy:O'),
    size = alt.Size('Exp_of_life_birth', scale = alt.Scale(zero = False)),
)

male

c:\Users\Thapakorn\AppData\Local\Programs\Python\Python310\lib\site-packages\altair\utils\core.py:317: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  for col_name, dtype in df.dtypes.iteritems():


alt.Chart(...)

In [ ]:
female

alt.Chart(...)

**The scaling is not the same, so the graph is combined in order to do a comparison between male and female**

For the comparison of the countries, looking at the male-only or female-only graph is enough for the comparison, because the size of the circles are very similar, but only the scaling is the problem.

In [ ]:
male + female

alt.LayerChart(...)

### Conclusion

&emsp;Overall, the expectation of life of every people around is increasing every year, but the increasing rate in each country is difference. The significant difference changes in expectation of life is Singapore female, which goes from 76.5 to 85.1 years old in 30 years, 8.6 years difference. Other than that, the improved expectation of life is around 2-4 years. 

&emsp;Comparing age-wise, in every country and every year, the expectation of life of the female is always higher than the male because the red circle area is bigger than the blue ones. The closet expectation of life between male and female is in Netherland in 2016 with 79.9 years old and 83.1 years old respectively. The biggest gap is France in 1986 with 71.5 years old for male and 79.7 years old for female, which is 8.2 years difference.

&emsp;The lowest expectation of life for male is U.S.A. in 1986 with only 71.2 years old, which is the lowest value in the dataset. On the other hand the highest expectation of life for male is Hong Kong in 2015 with 81.4 years old, 10.2 years difference from the lowest. For the female, the lowest is the Singapore in 1986 with 76.5 years old, while the highest is Hong Kong in 2015 and 2016 with 87.3 years old, 10.8 years difference from the lowest point.

&emsp;The countries that have the highest expectation of life are both Hong Kong and Japan, these two countries rival the expectation of life with each other since 1986 in both sexes. There are some period that Japan has longer lifetime than Hong Kong, and there are some period that Hong Kong has longer lifetime than Japan. However, the peak of the expectation of life is in Hong Kong which has been discussed earlier.

## 5. Number of death in each month (Seasonal death)

### Data retreival

In [ ]:
df = pd.read_excel(r"Mortality_trends.xlsx", sheet_name = 'Seasonality')

df

Yr    Month No_of_deaths Seasonal_ind
0         1986        1         2502            -
1         1986        2         2176            -
2         1986        3         2555            -
3         1986        4         2182            -
4         1986        5         2177            -
..         ...      ...          ...          ...
125  2006–2016        9            -           89
126  2006–2016       10            -           93
127  2006–2016       11            -           93
128  2006–2016       12            -          109
129  2006–2016  Average            -          100

[130 rows x 4 columns]

### Data cleaning/filtering

In [ ]:
# fliter only data that we used
df_hm = df[df['Yr'] != '2006–2016'][df['Yr'] != '1996–2006'][df['Yr'] != '1986–1996'][df['Month'] != '1–12'][['Yr', 'Month', 'No_of_deaths']]
df_hm[['Yr', 'Month']] = df_hm[['Yr', 'Month']].apply(pd.to_numeric)

df_hm

C:\Users\Thapakorn\AppData\Local\Temp\ipykernel_3396\798652009.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df_hm = df[df['Yr'] != '2006–2016'][df['Yr'] != '1996–2006'][df['Yr'] != '1986–1996'][df['Month'] != '1–12'][['Yr', 'Month', 'No_of_deaths']]


Yr  Month No_of_deaths
0   1986      1         2502
1   1986      2         2176
2   1986      3         2555
3   1986      4         2182
4   1986      5         2177
..   ...    ...          ...
85  2016      8         3535
86  2016      9         3375
87  2016     10         3602
88  2016     11         3643
89  2016     12         4006

[84 rows x 3 columns]

### Data visualization

In this topic, there are not many data to work on. The author want to show that there are many visualization types that can be used in order to exact the data that can be easily analyzed. This visualization is called heatmap.

In [ ]:
# using heatmap to plot
alt.Chart(df_hm).mark_rect().encode(
    x = 'Yr:N',
    y = 'Month:N',
    color = alt.Color('No_of_deaths:Q', scale = alt.Scale(scheme = 'greys'))
)

c:\Users\Thapakorn\AppData\Local\Programs\Python\Python310\lib\site-packages\altair\utils\core.py:317: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  for col_name, dtype in df.dtypes.iteritems():


alt.Chart(...)

### Conclusion

&emsp;Looking the graph horizontally, the number of deaths is increasing every year regardless of the month or the season. There are a few points that did not satisfy the aforementioned statement, such as March, 1996 the number of death is 3242, which is higher than the one in 1986, 1991, 2001, 2006. In fact, most of the value in 1996 will most likely have more value than the one in 2001, but the author used the March as an example because it is the most obvious one.

&emsp;Looking the graph vertically, the number of death will peak during the December to March, which is the winter season. However, the lowest point is occurred during the June to August, which is the summer season. With this information given, It can be concluded that winter is the most dangerous season, while the summer season is the safest season. In the author opinion, this is because there are many festivals or events happended during winter, which leads to more traveling, gathering, etc. and then leads to more accident. Thus, The opposite goes for the summer.

## Citation and References

Census and Statistics Department Hong Kong Special Administrative Region. "Demographic Trends in Hong Kong." Accessed December 12, 2022.
<br>
&emsp;&emsp;https://www.censtatd.gov.hk/en/EIndexbySubject.html?scode=150&pcode=B1120017#section3

Eurostat. "Glossary:Standardised death rate (SDR)." Accessed December 12, 2022.
<br> 
&emsp;&emsp;https://ec.europa.eu/eurostat/statistics-explained/index.php?title=Glossary:Standardised_death_rate_(SDR)